src/cunumeric/index/advanced_indexing.cu

```diff
@@ -0,0 +1,173 @@
+                       const Rect<DIM2>& rect,
+                       const size_t volume,
+                       cudaStream_t stream,
+                       Buffer<int64_t>& offsets) const
+  {
+    DeferredReduction<SumReduction<size_t>> size;
+
+    const size_t blocks = (volume + THREADS_PER_BLOCK - 1) / THREADS_PER_BLOCK;
+    size_t shmem_size   = THREADS_PER_BLOCK / 32 * sizeof(size_t);
+
+    if (blocks >= MAX_REDUCTION_CTAS) {
+      const size_t iters = (blocks + MAX_REDUCTION_CTAS - 1) / MAX_REDUCTION_CTAS;
+      count_nonzero_kernel<<<MAX_REDUCTION_CTAS, THREADS_PER_BLOCK, shmem_size, stream>>>(
+        volume, size, in, pitches, rect.lo, iters, offsets);
+    } else
+      count_nonzero_kernel<<<blocks, THREADS_PER_BLOCK, shmem_size, stream>>>(
+        volume, size, in, pitches, rect.lo, 1, offsets);
+
+    cudaStreamSynchronize(stream);
+
+    auto off_ptr = offsets.ptr(0);
+    thrust::exclusive_scan(thrust::cuda::par.on(stream), off_ptr, off_ptr + volume, off_ptr);
+
+    return size.read();
+  }
+
+  template <typename OUT_TYPE>
+  size_t operator()(Buffer<OUT_TYPE>& out,
+                    const AccessorRO<VAL, DIM1>& input,
+                    const AccessorRO<bool, DIM2>& index,
+                    const Pitches<DIM1 - 1>& pitches_input,
+                    const Rect<DIM1>& rect_input,
+                    const Pitches<DIM2 - 1>& pitches_index,
+                    const Rect<DIM2>& rect_index) const
+  {
+#ifdef DEBUG_CUNUMERIC
+    // in this case shapes for input and index arrays  should be the same
+    assert(Domain(rect_input) == Domain(rect_index));
+#endif
+    size_t size           = 0;
+    const bool* index_ptr = index.ptr(rect_index);
+    const size_t volume   = rect_index.volume();
+    cudaStream_t stream;
+    cudaStreamCreate(&stream);
+    auto offsets = create_buffer<int64_t>(volume, Memory::Kind::GPU_FB_MEM);
+    size         = compute_size(index, pitches_index, rect_index, volume, stream, offsets);
+
+    out = create_buffer<OUT_TYPE>(size, Memory::Kind::GPU_FB_MEM);
+    // populate output
+    if (size > 0) {
+      const size_t blocks = (volume + THREADS_PER_BLOCK - 1) / THREADS_PER_BLOCK;
+      advanced_indexing_kernel<<<blocks, THREADS_PER_BLOCK, 0, stream>>>(volume,
+                                                                         input,
+                                                                         index,
+                                                                         out,
+                                                                         pitches_input,
+                                                                         rect_input.lo,
+                                                                         pitches_index,
+                                                                         rect_index.lo,
+                                                                         offsets);
+    }
+    return size;
+  }
+};
+
```

Comment:
Please follow the GPU task debugging regime established in a [recent PR](https://github.com/nv-legate/cunumeric/pull/281):

* Use `Buffer` and `create_buffer` instead of `DeferredBuffer`.
* Don't create custom streams, instead call `get_cached_stream`.
* Use the above stream on all kernel launches.
* Call `CHECK_CUDA_STREAM(stream)` after every kernel launch.